File: go/name_utils.go
package modal

import (
	"fmt"
	"regexp"
)

var (
	objectNameCharset = regexp.MustCompile(`^[a-zA-Z0-9\-_.]+$`)
	appIDPattern      = regexp.MustCompile(`^ap-[a-zA-Z0-9]{22}$`)
)

// isValidObjectName checks whether a name is a valid Modal object name.
func isValidObjectName(name string) bool {
	return len(name) <= 64 &&
		objectNameCharset.MatchString(name) &&
		!appIDPattern.MatchString(name)
}

// checkObjectName validates a Modal object name, returning an InvalidError
// if the name is not valid. The objectType is used in the error message
// (e.g. "Image", "Image tag").
func checkObjectName(name string, objectType string) error {
	if !isValidObjectName(name) {
		return InvalidError{
			Exception: fmt.Sprintf(
				"Invalid %s name: '%s'."+
					"\n\nNames may contain only alphanumeric characters, dashes, periods, and underscores,"+
					" must be shorter than 64 characters, and cannot conflict with App ID strings.",
				objectType, name,
			),
		}
	}
	return nil
}

File: js/src/name_utils.ts
import { InvalidError } from "./errors";

const OBJECT_NAME_CHARSET = /^[a-zA-Z0-9\-_.]+$/;
const APP_ID_PATTERN = /^ap-[a-zA-Z0-9]{22}$/;

/**
 * Check whether a name is a valid Modal object name.
 *
 * @internal
 * @hidden
 */
export function isValidObjectName(name: string): boolean {
  return (
    name.length <= 64 &&
    OBJECT_NAME_CHARSET.test(name) &&
    !APP_ID_PATTERN.test(name)
  );
}

/**
 * Validate a Modal object name, throwing an {@link InvalidError} if invalid.
 * The `objectType` is used in the error message (e.g. "Image", "Image tag").
 *
 * @internal
 * @hidden
 */
export function checkObjectName(name: string, objectType: string): void {
  if (!isValidObjectName(name)) {
    throw new InvalidError(
      `Invalid ${objectType} name: '${name}'.` +
        "\n\nNames may contain only alphanumeric characters, dashes, periods, and underscores," +
        " must be shorter than 64 characters, and cannot conflict with App ID strings.",
    );
  }
}
